src/modules/student/application/student.service.ts

```diff
@@ -57,4 +70,97 @@ export class StudentService {
     async findAll(): Promise<Student[]> {
         return await this.studentRepository.findAll();
     }
+
+    async update(id: number, dto: UpdateStudentDTO){
+        const existingStudent = await this.studentRepository.findById(id);
+        if (!existingStudent) {
+            throw new NotFoundException(`Student with ID ${id} not found.`);
+        }
+
+        if (dto.registrationNumber) {
+            if (!UpdateStudentDTO.validateCPF(dto.registrationNumber)) {
+                throw new BadRequestException("Invalid CPF");
+            }
+            const regOwner = await this.studentRepository.findByRegistrationNumber(dto.registrationNumber);
+            if (regOwner && regOwner.getId() !== id) {
+                throw new ConflictException(`CPF '${dto.registrationNumber}' is already in use.`);
+            }
+        }
+
+         if (dto.dateOfBirth) {
+            if (dto.dateOfBirth > new Date()) {
+                throw new BadRequestException('Date of birth cannot be in the future.');
+            }
+        }
+
+
+        Object.keys(dto).forEach(key => {
+            const setterName = `set${key.charAt(0).toUpperCase() + key.slice(1)}`;
+            if (typeof existingStudent[setterName] === 'function') {
+                existingStudent[setterName](dto[key]);
+            }
+        });
+
+        return await this.studentRepository.update(existingStudent);
+    }
+
+    async delete(id: number): Promise<void> {
+        await this.findById(id);
+        return this.studentRepository.delete(id);
+    }
+
+    async addAddressToStudent(studentId: number, dto: CreateAddressDTO): Promise<AddressEntity> {
+        const student = await this.findById(studentId);
+        if (!student) {
+            throw new NotFoundException(`Student with ID ${studentId} not found.`);
+        }
```

Comment:
```suggestion
        }
        if(student.getAddressId()){
            throw new ConflictException(`Address assignment failed: Student ${studentId} already has an associated address. To change it, update the existing address or remove it first.`);
        }
```
Antes de inserir um novo endereço, é necessário testar se já não havia um anteriormente.